extensions/amp-bind/0.1/bind-expression.js

```diff
@@ -14,33 +14,253 @@
+        String.prototype.split,
+        String.prototype.substr,
+        String.prototype.substring,
+        String.prototype.toLowerCase,
+        String.prototype.toUpperCase,
+      ],
+  };
+  // Creates a prototype-less map of function name to the function itself.
+  // This makes function lookups faster (compared to Array.indexOf).
+  const out = Object.create(null);
+  Object.keys(whitelist).forEach(type => {
+    out[type] = Object.create(null);
+
+    const functions = whitelist[type];
+    for (let i = 0; i < functions.length; i++) {
+      const f = functions[i];
+      out[type][f.name] = f;
+    }
+  });
+  return out;
+})();
+
+/**
  * A single Bind expression.
  */
 export class BindExpression {
   /**
    * @param {string} expressionString
+   * @throws {Error} On malformed expressions.
    */
   constructor(expressionString) {
     /** @const {string} */
     this.expressionString = expressionString;
+
+    /** {!./bind-expr-defines.AstNode} */
+    this.ast_ = parser.parse(this.expressionString);
   }
 
   /**
    * Evaluates the expression given a scope.
    * @param {!Object} scope
+   * @throws {Error} On illegal function invocation.
    * @return {*}
    */
   evaluate(scope) {
-    // TODO(choumx): Improve performance by extracting AST construction
-    // and generating evaluator with Function constructor.
-    try {
-      parser.yy = scope;
-      return parser.parse(this.expressionString);
-    } finally {
-      parser.yy = null;
+    return this.eval_(this.ast_, scope);
+  }
+
+  /**
+   * Recursively evaluates and returns value of `node` and its children.
+   * @param {?./bind-expr-defines.AstNode} node
+   * @param {!Object} scope
+   * @throws {Error}
+   * @return {*}
+   * @private
+   */
+  eval_(node, scope) {
+    if (!node) {
+      return null;
+    }
+
+    const {type, args, value} = node;
+
+    if (type === AstNodeType.LITERAL) {
+      return value;
+    }
+
+    switch (type) {
+      case AstNodeType.EXPRESSION:
+        return this.eval_(args[0], scope);
+
+      case AstNodeType.INVOCATION:
+        const caller = this.eval_(args[0], scope);
+        const params = this.eval_(args[1], scope);
+        const method = String(value);
+
+        const callerType = Object.prototype.toString.call(caller);
+        const whitelist = FUNCTION_WHITELIST[callerType];
+        if (whitelist) {
+          const func = caller[method];
+          if (func && func === whitelist[method]) {
+            if (Array.isArray(params) && !this.containsObject_(params)) {
+              return func.apply(caller, params);
+            } else {
+              throw new Error(`Unexpected argument type in ${method}().`);
+            }
+          }
+        }
+        throw new Error(`${method}() is not a supported function.`);
+
+      case AstNodeType.MEMBER_ACCESS:
+        const target = this.eval_(args[0], scope);
+        const member = this.eval_(args[1], scope);
+
+        if (target === null || member === null) {
```

Comment:
Soft-failure for undefined variable errors is by design. Angular does something similar.